### ISSUE TITLE
Webpack 4: Cannot resolve dependancy of symlinked module

### ISSUE THREAD

**Issue by omar-bibliocommons** [webpack/webpack#8824]
Here's a demo repo which replicates the issue being discussed:

https://bitbucket.org/omar-biblio/webpack-symlink-error-demo/src/master/

After upgrading from Webpack 1 to 4, I've encountered the following error:

```ERROR in ./node_modules/@my_symlinked_packages/base-accessibility-date-picker/lib/components/AcessibleDatePicker.component.js
Module not found: Error: Can't resolve 'react-dates/lib/utils/isInclusivelyAfterDay' in '/path_to_my_local_app/node_modules/@my_symlinked_packages/base-accessibility-date-picker/lib/components'
 @ ./node_modules/@my_symlinked_packages/base-accessibility-date-picker/lib/components/AcessibleDatePicker.component.js 32:29-83
 @ ./node_modules/@my_symlinked_packages/base-accessibility-date-picker/lib/index.js
 @ ./app/components/borrowing/holds/HoldsPauseDateSelect/HoldsPauseDateSelect.jsx
 @ ./app/components/borrowing/holds/HoldsPauseDateSelect/index.js
 @ ./app/components/borrowing/holds/PauseHoldWorkflowView/components/BasePauseHoldWorkflowView.jsx
 @ ./app/components/borrowing/holds/PauseHoldWorkflowView/components/SinglePauseHoldWorkflowView.jsx
 @ ./app/containers/ItemTransactionWorkflowContainer.jsx
 @ ./app/containers/WorkflowContainer.jsx
 @ ./app/components/widgets/ItemWidget/ItemWidget.jsx
 @ ./app/components/widgets/ItemWidget/index.js
 @ ./app/containers/widgets/ItemWidgetContainer.jsx
 @ ./app/routes/Routes.jsx
 @ ./app/bootstrap.js
 @ multi webpack-dev-server/client?http://localhost:3002 webpack/hot/only-dev-server ./app/bootstrap.js
```

I believe this error is indicating that Wepback 4 cannot resolve react-dates module which is a dependancy of a module that I am symlinking.

Here's a sample of my project layout:

```
my_symlinked_packages
|-node_modules
|-package.json
|-src
  |-my_lerna_package1
    |-lib
    |-specs
    |-src
    |-package.json (contains react-dates as a dependancy)


consuming_application
|-node_modules
  |-@my_symlinked_packages
    |-my_lerna_package1
|-package.json
|-webpack.config.js
|-app
```

Here's an example of my webpack.config.js (it's being composed of several js imports)

```
 //import dependancies
  //assign value to custom variables

  module.exports = { 
    devtool: 'eval',
    context: '/path_to_my_local_app',
    mode: 'development',
    entry:
    { app:
        [ 'webpack-dev-server/client?http://localhost:3002',
          'webpack/hot/only-dev-server',
          './app/bootstrap.js' ],
      vendors:
        [ 'raf/polyfill',
          'babel-polyfill',
          'immutable',
          'iso',
          'jquery',
          'lodash',
          'moment',
          'react',
          'react-router',
          'redux',
          'react-redux',
          'superagent',
          'when',
          'rxjs',
          'handlebars' ] },
    externals: [ 'farmhash' ],
    output:
    { path: '/path_to_my_local_app/public',
      filename: '[name].js',
      chunkFilename: '[name].js',
      publicPath: 'http://localhost:3002/' },
    module:
    { rules:
        [ {
          use: ['react-hot-loader/webpack', 'happypack/loader?id=babel'],
          test:    /\.jsx?$/,
          exclude: /node_modules/,
          include: /app|server|node_modules\/@my_symlinked_packages/
        },
        {
          test: /\.jsx?$/,
          loader: 'eslint-loader',
          exclude: /node_modules/,
          include: /app|server/
        },
        {
          test:   /\.s?css$/,
          use: [
            'style-loader',
            {
              loader:'css-loader',
              options: {
                importLoaders: 2,
                sourceMap: true
              }
            },
            { 
              loader: 'postcss-loader',
              options: { sourceMap: true, config: { path: postcssConfig } }
            },
            {
              loader: 'sass-loader',
              options: {
                outputStyle: 'expanded',
                sourceMap: true
              }
            }]
        },
        {
          test:   /\.(png|jpg|gif|ico|woff|woff2|eot|ttf|svg)$/,
          loader: 'url-loader?limit=' + inlinedAssetSizeLimit
        },
        {
          test: require.resolve('jquery'),
          loader: 'expose-loader?$!expose-loader?jQuery'
        },
        {
          test: /handlebars\.js/,
          loader: 'expose-loader?Handlebars'
        } ] },
    resolve:
    { symlinks: false,
      modules:
        [ '/path_to_my_local_app/app',
          '/path_to_my_local_app/node_modules',
          '/path_to_my_local_app/vendor' ],
      alias:
        { handlebars: 'handlebars/dist/handlebars.js',
          containers: '/path_to_my_local_app/app/containers',
          constants: '/path_to_my_local_app/app/constants',
          actions: '/path_to_my_local_app/app/actions' },
      extensions: [ '.js', '.jsx' ] },
    plugins: [ // hot reload
      new HappyPack({
        id: 'babel',
        threads: 4,
        loaders: [
          {
            loader:  'babel-loader',
            query: {
              cacheDirectory: true
            }
          }
        ]
      }),
      new webpack.HotModuleReplacementPlugin(),
      new webpack.DefinePlugin({
        __CLIENT__: true,
        __SERVER__: false,
        __DEVELOPMENT__: true,
        __DISABLE_SSR__: false
      }),
      new webpack.LoaderOptionsPlugin({
        options: {
          sassLoader: {
            data: `$is-widget: ${widgetBuild ? 'true' : 'false'};`
          },
          eslint: {
            emitWarning: true
          },
          progress: true
        }
      }),

      // https://github.com/halt-hammerzeit/webpack-isomorphic-tools
      new WebpackIsomorphicToolsPlugin(webpackIsomorphicToolsConfig).development()
    ]
  }
```
And here is a snippet of my postcss.config.js

```
module.exports = {
  plugins: {
    autoprefixer: {
      browsers: ['last 2 version', 'ie >= 9', 'iOS >= 8']
    }
  }
};
```

Any help would be welcome!

**Comment by dbslone** [comment]
I think I am running into the same issue. I'm trying to use webpack alias with webpack-middlewares for express. I have created a generic repo here https://github.com/dbslone/webpack-alias-playground that reproduces the problem.

**Comment by omar-bibliocommons** [comment]
@dbslone I cloned your repo and the issue that I was able to reproduce was: 

```
ERROR in ./src/js/Message.js
Module not found: Error: Can't resolve 'analytics' in '/Users/omarplummer/Documents/webpack-alias-playground/src/js'
 @ ./src/js/Message.js 11:17-37
 @ ./src/index.js
 @ multi webpack-hot-middleware/client?path=/__webpack_hmr&timeout=20000 ./src/index.js
```

I was able to fix this issue by correctly referencing the analytics file in src/js/Message.js

```
//i updated the import with the following path:
import Analytics from '../analytics'
```

**Comment by omar-bibliocommons** [comment]
My current work around to my original issue is to install the missing dependancy in the consuming application but this is not an optimal solution as I'm expecting all dependancies from my symlinked lerna mono repos to be resolved correctly

**Comment by sokra** [comment]
For `resolve.symlinks: true` or default: The dependency must be able to be resolved from the real location of the symlinked file.

For `resolve.symlinks: false`: The dependency must be able to be resolved from the linked location of the symlinked file.

**Comment by omar-bibliocommons** [comment]
@sokra I've tried modifying the resolve.symlinks option from false to true and I still have the same issue. Here's a link to a demo repo which replicates the issue I"m having with my application for anyone would like like to try taking a look at the issue:

https://bitbucket.org/omar-biblio/webpack-symlink-error-demo/src/master/

**Comment by omar-bibliocommons** [comment]
We were able to solve this issue by adding a resolve path pointing to the nodule modules of the symlinked directory.

```
// example shows resolve config pointing to the directory which stores the
// symlinked mono repos
    module.exports = {
      resolve : {
        modules: [
          path.resolve('./app'),
          path.resolve('./node_modules'),
          path.resolve('./vendor'),
          path.resolve('../my_symlinked_packages/node_modules')
        ]
      }
    };
```

**Comment by bangolin** [comment]
Also make sure you have bundle and yarn installed and executed in the linked package

**Comment by Sly1024** [comment]
> @sokra I've tried modifying the resolve.symlinks option from false to true and I still have the same issue. Here's a link to a demo repo which replicates the issue I"m having with my application for anyone would like like to try taking a look at the issue:
> 
> https://bitbucket.org/omar-biblio/webpack-symlink-error-demo/src/master/

The default of `resolve.symlinks` is true, so you need to change it to `false`. That worked for me.

**Comment by lostpebble** [comment]
This seems to be a bit of a headache when it comes to monorepos and trying to balance linking external modules through something like `yarn link` / `npm link`, and also keeping your local symlinked packages working correctly with HMR and other things.

@sokra could there ever be a consideration for conditionally controlling the symlink option? I'd like any modules which originate in my project files to be `symlink: true` so that they continue to work properly within the context of my project as regular source files, and any other external modules linked in to be `symlink: false` so that they are disguised as being a part of my project and make use of the local installed dependencies.

I've recently discovered that I can't get things like HMR to work in my monorepo unless I set my config to `symlink: true` - but this brings the headache of now not being able to link external modules anymore without getting those dependency errors.
